lib/Backend/Lower.cpp

```diff
@@ -17931,6 +17934,70 @@ Lowerer::GenerateFastInlineArrayPop(IR::Instr * instr)
     GenerateHelperToArrayPopFastPath(instr, doneLabel, bailOutLabelHelper);
 }
 
+void
+Lowerer::GenerateFastInlineHasOwnProperty(IR::Instr * instr)
+{
+    Assert(instr->m_opcode == Js::OpCode::CallDirect);
+
+    //CallDirect src2
+    IR::Opnd * linkOpnd = instr->GetSrc2();
+    //ArgOut_A_InlineSpecialized
+    IR::Instr * tmpInstr = linkOpnd->AsSymOpnd()->m_sym->AsStackSym()->m_instrDef;
+
+    IR::Opnd * argsOpnd[2] = { 0 };
+    bool result = instr->FetchOperands(argsOpnd, 2);
+    Assert(result);
+    AnalysisAssert(argsOpnd[0] && argsOpnd[1]);
+
+    // fast path case where hasOwnProperty is being called using a property name loaded via a for-in loop
+    if (!argsOpnd[1]->GetValueType().IsString()
+        || argsOpnd[0]->GetValueType().IsNotObject()
+        || !argsOpnd[0]->IsRegOpnd()
+        || !argsOpnd[1]->IsRegOpnd()
+        || !argsOpnd[1]->AsRegOpnd()->m_sym->m_isSingleDef
+        || (argsOpnd[1]->AsRegOpnd()->m_sym->m_instrDef->m_opcode != Js::OpCode::BrOnEmpty
+            && argsOpnd[1]->AsRegOpnd()->m_sym->m_instrDef->m_opcode != Js::OpCode::BrOnNotEmpty))
+    {
+        return;
+    }
+
+    IR::LabelInstr *doneLabel = InsertLabel(false, instr->m_next);
+    IR::LabelInstr *labelHelper = InsertLabel(true, instr);
+    IR::Instr * insertInstr = labelHelper;
+    IR::RegOpnd * thisObj = argsOpnd[0]->AsRegOpnd();
+    IR::Opnd * forInEnumeratorOpnd = argsOpnd[1]->AsRegOpnd()->m_sym->m_instrDef->GetSrc1();
+
+    // go to helper if we can't use JIT fastpath
+    IR::Opnd * canUseJitFastPathOpnd = GetForInEnumeratorFieldOpnd(forInEnumeratorOpnd, Js::ForInObjectEnumerator::GetOffsetOfCanUseJitFastPath(), TyInt8);
```

Comment:
Please add a comment about what checking this flag will rule out (e.g. the for in object doesn't have any enumerable prototype property, it doesn't have any index property, the type has to be "lockable" (e.g. currently no DictionaryTypeHandler).  Also add a comment in the ForInObjectEnumerator for this flag so that if it ever expand the cases that we can use jit fast path, we will have to look at this fast path as well.